File: controller/adminController.js
const { prisma } = require('../prisma/client');

async function getAdminStatsController(req, res) {
  try {
    const thirtyDaysAgo = new Date(new Date().setDate(new Date().getDate() - 30));

    const [users, products, orders, revenueAgg] = await Promise.all([
      prisma.user.count(),
      prisma.product.count(),
      prisma.order.count(),
      prisma.order.aggregate({ _sum: { totalPrice: true } }),
    ]);

    const dailyUsers = await prisma.user.groupBy({
      by: ['createdAt'],
      where: { createdAt: { gte: thirtyDaysAgo } },
      _count: { id: true },
      orderBy: { createdAt: 'asc' },
    });

    const dailyOrders = await prisma.order.groupBy({
      by: ['createdAt'],
      where: { createdAt: { gte: thirtyDaysAgo } },
      _count: { id: true },
      _sum: { totalPrice: true },
      orderBy: { createdAt: 'asc' },
    });

    const top = await prisma.orderItem.groupBy({
      by: ['productVariantId'],
      _sum: { quantity: true },
      orderBy: { _sum: { quantity: 'desc' } },
      take: 6,
    });
    const variantIds = top.map(function(t){ return t.productVariantId; });
    const variants = variantIds.length ? await prisma.productVariant.findMany({
      where: { id: { in: variantIds } },
      include: { product: { include: { category: true } } },
    }) : [];
    const topProducts = variants.map(function(v){ return v.product; });

    function formatDailyData(data, valueField, countField) {
      const result = {};
      for (let i = 0; i < 30; i++) {
        const d = new Date(new Date().setDate(new Date().getDate() - i));
        result[d.toISOString().split('T')[0]] = 0;
      }
      data.forEach(function(d){
        const date = d.createdAt.toISOString().split('T')[0];
        result[date] = valueField ? d._sum[valueField] : d._count[countField];
      });
      return Object.entries(result)
        .sort(function([d1],[d2]){ return new Date(d1) > new Date(d2) ? 1 : -1; })
        .reduce(function(acc,[d,v]){ acc[d]=v; return acc; },{});
    }

    res.status(200).json({
      users,
      products,
      orders,
      revenue: revenueAgg._sum.totalPrice,
      topProducts,
      charts: {
        users: formatDailyData(dailyUsers, null, 'id'),
        orders: formatDailyData(dailyOrders, null, 'id'),
        revenue: formatDailyData(dailyOrders, 'totalPrice', null),
      },
    });
  } catch (error) {
    console.error('Admin Stats Error:', error);
    res.status(500).json({ message: error.message });
  }
}

async function listUsersController(req, res) {
  try {
    const users = await prisma.user.findMany({ select: { id:true, firstName:true, lastName:true, email:true, role:true, createdAt:true } });
    res.status(200).json(users);
  } catch (error) { res.status(500).json({ message: error.message }); }
}

async function listOrdersController(req, res) {
  try {
    const orders = await prisma.order.findMany({ include: { user: { select: { firstName:true, lastName:true } }, items: true } });
    res.status(200).json(orders);
  } catch (error) { res.status(500).json({ message: error.message }); }
}

module.exports = { getAdminStatsController, listUsersController, listOrdersController };

File: Backend/Controllers/AdminController.cs
using AllureModa.API.Data;
using AllureModa.API.DTOs.Admin;
using AllureModa.API.Models;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;

namespace AllureModa.API.Controllers
{
    [ApiController]
    [Route("api/admin")]
    [Authorize(Roles = "ADMIN")]
    public class AdminController : ControllerBase
    {
        private readonly ApplicationDbContext _context;

        public AdminController(ApplicationDbContext context)
        {
            _context = context;
        }

        [HttpGet("stats")]
        public async Task<ActionResult<AdminStatsResponse>> GetStats()
        {
            var thirtyDaysAgo = DateTime.UtcNow.AddDays(-30);

            // 1. Basic Counts (Parallel execution if possible, EF Core allows async mostly)
            // 1. Basic Counts (Execute sequentially to avoid DbContext concurrency issues)
            var usersCount = await _context.Users.CountAsync();
            var productsCount = await _context.Products.CountAsync();
            var ordersCount = await _context.Orders.CountAsync();
            var revenue = await _context.Orders.SumAsync(o => o.TotalPrice);

            // 2. Charts Data (Group By Date) - Optimized in DB
            // Note: Grouping by Date in EF Core with PostgreSQL can be tricky with timezones.
            // Using basic Truncate or explicit Date property if available. 
            // For now, we fetch minimal data or try EF.Functions.
            
            var dailyUsers = await _context.Users
                .Where(u => u.CreatedAt >= thirtyDaysAgo)
                .GroupBy(u => u.CreatedAt.Date)
                .Select(g => new { Date = g.Key, Count = g.Count() })
                .ToListAsync();

            var dailyOrders = await _context.Orders
                .Where(o => o.CreatedAt >= thirtyDaysAgo)
                .GroupBy(o => o.CreatedAt.Date)
                .Select(g => new { Date = g.Key, Count = g.Count(), Revenue = g.Sum(x => x.TotalPrice) })
                .ToListAsync();

            // 3. Top Products
            // Needs to join OrderItems -> ProductVariant -> Product
            var topVariants = await _context.OrderItems
                .GroupBy(oi => oi.ProductVariantId)
                .Select(g => new { VariantId = g.Key, Quantity = g.Sum(oi => oi.Quantity) })
                .OrderByDescending(x => x.Quantity)
                .Take(6)
                .ToListAsync();

            var variantIds = topVariants.Select(x => x.VariantId).ToList();
            
            var topProducts = await _context.ProductVariants
                .Where(v => variantIds.Contains(v.Id))
                .Select(v => v.Product!) // Select the parent product
                .Distinct()
                .Include(p => p.Category)
                .ToListAsync();

            // Format Charts
            var charts = new DashboardCharts();
            // Fill with 0 for missing days (Logic similar to Node)
            for (int i = 0; i < 30; i++)
            {
                var date = DateTime.UtcNow.AddDays(-i).Date.ToString("yyyy-MM-dd");
                charts.Users[date] = 0;
                charts.Orders[date] = 0;
                charts.Revenue[date] = 0;
            }

            foreach (var item in dailyUsers)
            {
                var date = item.Date.ToString("yyyy-MM-dd");
                if (charts.Users.ContainsKey(date)) charts.Users[date] = item.Count;
            }

            foreach (var item in dailyOrders)
            {
                var date = item.Date.ToString("yyyy-MM-dd");
                if (charts.Orders.ContainsKey(date))
                {
                    charts.Orders[date] = item.Count;
                    charts.Revenue[date] = item.Revenue;
                }
            }

            return Ok(new AdminStatsResponse
            {
                Users = usersCount,
                Products = productsCount,
                Orders = ordersCount,
                Revenue = revenue,
                TopProducts = topProducts,
                Charts = charts
            });
        }

        [HttpGet("users")]
        public async Task<ActionResult<IEnumerable<User>>> ListUsers([FromQuery] int page = 1, [FromQuery] int limit = 50)
        {
            var users = await _context.Users
                .OrderByDescending(u => u.CreatedAt)
                .Skip((page - 1) * limit)
                .Take(limit)
                .ToListAsync();
            
            return Ok(users);
        }

        [HttpGet("orders")]
        public async Task<ActionResult<IEnumerable<Order>>> ListOrders([FromQuery] int page = 1, [FromQuery] int limit = 50)
        {
            var orders = await _context.Orders
                .Include(o => o.User)
                .Include(o => o.Items)
                .OrderByDescending(o => o.CreatedAt)
                .Skip((page - 1) * limit)
                .Take(limit)
                .ToListAsync();

            return Ok(orders);
        }
    }
}
